prov/sm2/src/sm2_ep.c

```diff
@@ -519,197 +384,13 @@ static int sm2_ep_bind(struct fid *ep_fid, struct fid *bfid, uint64_t flags)
-		       uint64_t tag, struct fi_peer_rx_entry **rx_entry)
-{
-	struct sm2_rx_entry *sm2_entry;
-	struct sm2_srx_ctx *srx_ctx;
-	struct sm2_match_attr match_attr;
-	struct dlist_entry *dlist_entry;
-	int ret;
-
-	srx_ctx = srx->ep_fid.fid.context;
-	ofi_spin_lock(&srx_ctx->lock);
-
-	match_attr.id = addr;
-	match_attr.tag = tag;
-
-	dlist_entry = dlist_find_first_match(&srx_ctx->trecv_queue.list,
-					     srx_ctx->trecv_queue.match_func,
-					     &match_attr);
-	if (!dlist_entry) {
-		sm2_entry = sm2_alloc_rx_entry(srx_ctx);
-		if (!sm2_entry) {
-			ret = -FI_ENOMEM;
-		} else {
-			sm2_entry->peer_entry.owner_context = NULL;
-			sm2_entry->peer_entry.addr = addr;
-			sm2_entry->peer_entry.size = size;
-			sm2_entry->peer_entry.tag = tag;
-			sm2_entry->peer_entry.srx = srx;
-			*rx_entry = &sm2_entry->peer_entry;
-			ret = -FI_ENOENT;
-		}
-		goto out;
-	}
-	dlist_remove(dlist_entry);
-
-	*rx_entry = (struct fi_peer_rx_entry *) dlist_entry;
-	(*rx_entry)->srx = srx;
-	ret = FI_SUCCESS;
-out:
-	ofi_spin_unlock(&srx_ctx->lock);
-	return ret;
-}
-
-static int sm2_queue_msg(struct fi_peer_rx_entry *rx_entry)
-{
-	struct sm2_srx_ctx *srx_ctx = rx_entry->srx->ep_fid.fid.context;
-
-	ofi_spin_lock(&srx_ctx->lock);
-	dlist_insert_tail((struct dlist_entry *) rx_entry,
-			  &srx_ctx->unexp_msg_queue.list);
-	ofi_spin_unlock(&srx_ctx->lock);
-	return 0;
-}
-
-static int sm2_queue_tag(struct fi_peer_rx_entry *rx_entry)
-{
-	struct sm2_srx_ctx *srx_ctx = rx_entry->srx->ep_fid.fid.context;
-
-	ofi_spin_lock(&srx_ctx->lock);
-	dlist_insert_tail((struct dlist_entry *) rx_entry,
-			  &srx_ctx->unexp_tagged_queue.list);
-	ofi_spin_unlock(&srx_ctx->lock);
-	return 0;
-}
-
-static void sm2_free_entry(struct fi_peer_rx_entry *entry)
-{
-	struct sm2_srx_ctx *srx =
-		(struct sm2_srx_ctx *) entry->srx->ep_fid.fid.context;
-	struct sm2_rx_entry *sm2_entry, *owner_entry;
-
-	ofi_spin_lock(&srx->lock);
-	sm2_entry = container_of(entry, struct sm2_rx_entry, peer_entry);
-	if (entry->owner_context) {
-		owner_entry = container_of(entry->owner_context,
-					   struct sm2_rx_entry, peer_entry);
-		if (!--owner_entry->multi_recv_ref &&
-		    owner_entry->peer_entry.size < srx->min_multi_recv_size) {
-			if (ofi_peer_cq_write(srx->cq,
-					      owner_entry->peer_entry.context,
-					      FI_MULTI_RECV, 0, NULL, 0, 0,
-					      FI_ADDR_NOTAVAIL)) {
-				FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
-					"unable to write rx MULTI_RECV "
-					"completion\n");
-			}
-			ofi_freestack_push(srx->recv_fs, owner_entry);
-		}
-	}
-
-	ofi_freestack_push(srx->recv_fs, sm2_entry);
-	ofi_spin_unlock(&srx->lock);
-}
-
-static struct fi_ops_srx_owner sm2_srx_owner_ops = {
-	.size = sizeof(struct fi_ops_srx_owner),
-	.get_msg = sm2_get_msg,
-	.get_tag = sm2_get_tag,
-	.queue_msg = sm2_queue_msg,
-	.queue_tag = sm2_queue_tag,
-	.free_entry = sm2_free_entry,
-};
-
 static int sm2_discard(struct fi_peer_rx_entry *rx_entry)
 {
 	struct sm2_xfer_ctx *xfer_ctx = rx_entry->peer_context;
```

Comment:
@a-szegel I noticed a bug here and started to fix it but then realized I have to do protocol things to return the command to the sender. I'm going to let you fix that. Interesting that no testing caught this though. Might point to something missing in cleanup. Something to double check.